content/post/ddos-detection.md

```diff
@@ -0,0 +1,413 @@
+author = "nisanharamati"
++++
+
+
+# DDoS Attack Detection with Wallaroo: A Real-time Time Series Analysis Example
+
+In this post, I would like to demonstrate a real-world use case for Wallaroo: DDoS attack detection in real time, using a stream of request logs from a large group of servers. This example builds on the concepts introduced in [Non-native event-driven windowing in Wallaroo][windowing in wallaroo], adding partitioning, parallelization, and the application of statistical analysis to perform a simplified anomaly detection based on weighted mean and standard deviation to determine when a server is experiencing an uncharacteristic and heavy load.
+
+If you're new to Wallaroo and are interested in learning more about it before diving into today's example, you can check our [Go Python, Go! Stream Processing for Python][go python] post, where we introduce the Wallaroo Python API.
+
+The code for the example in this post can be found on [GitHub][demo code], and can be run with the [current version of Wallaroo (0.2.2)](https://github.com/WallarooLabs/wallaroo/tree/0.2.2).
+
+## Your Service Layer is Under Attack
+
+Running a large application over the internet is really great! You can reach a massive customer base, and provide them with your digital product in no time at all. Unless...
+
+![503][503 service unavailable]
+(source: https://andrewodendaal.com/503-service-unavailable-back-end-server-is-at-capacity/)
+![502 bad gateway][502 bad gateway]
+![this site cannot be reached][this site cannot be reached]
+
+Well that sucks.
+
+There are a lot of reasons why a service can become unavailable or unreachable, and many of them are benign, like hardware hiccups, or bad code, or maybe a messed up routing table; one reason, however, is not benign: a denial of service (DoS) attack, or more commonly these days, a distributed denial of service (DDoS) attack.
+
+According to [Cloudflare][cloudflare what is ddos],
+
+> [a] distributed denial-of-service (DDoS) attack is a malicious attempt to disrupt normal traffic of a targeted server, service or network by overwhelming the target or its surrounding infrastructure with a flood of Internet traffic.
+
+In other words, the attack utilizes the fact that responding to each _request_ takes up a fraction of your system's finite resources, so if an attacker can generate sufficiently many requests, they can use up _all_ of your resources, leaving none available to service legitimate requests from your users. But that seems easy to defend against, right? Just throttle traffic by user or session! Ah, but this is where the "distributed" part comes in. In order to avoid being throttled, an attacker will use a large number of (probably) compromised machines to send service requests. At that point, it can become difficult or expensive to determine which new request from an unknown user may be legitimate and which may be malicious.
```

Comment:
"In other words"

drop that. provides no value

Reply:
The entire "in other words" sentence is really long. lots of clauses. There are several in this post like that. I have a hard time following. Breaking your clauses up into shorter sentences that are more compact would help me follow along.

For example

", so if an attacker"

There's no reason for that to be a clause. its a new sentence.